Sources/apollo-ios-pagination/GraphQLQueryPager.swift

```diff
@@ -0,0 +1,212 @@
+  ///   - onError: The callback when there is an error.
+  public init<P: PaginationInfo>(
+    client: ApolloClientProtocol,
+    initialQuery: InitialQuery,
+    extractPageInfo: @escaping (PageExtractionData) -> P,
+    nextPageResolver: @escaping (P) -> PaginatedQuery
+  ) {
+    self.client = client
+    self.initialQuery = initialQuery
+    self.extractPageInfo = extractPageInfo
+    self.nextPageResolver = { page in
+      guard let page = page as? P else { return nil }
+      return nextPageResolver(page)
+    }
+  }
+
+  // MARK: - Public API
+
+  /// Subscribe to new data from this watcher.
+  /// - Returns: An async stream that can be iterated over.
+  public func subscribe() -> AsyncStream<Result<Output, Error>> {
+    AsyncStream { continuation in
+      self.onUpdate = { continuation.yield(.success($0)) }
+      self.onError = { continuation.yield(.failure($0)) }
+      continuation.onTermination = { @Sendable [weak self] _ in
+        self?.cancel()
+      }
+    }
+  }
+
+  /// Subscribe to new data from this watcher, using a callback
+  /// - Parameter onUpdate: A callback function that supplies a `Result<Output, Error>`
+  public func subscribe(onUpdate: @MainActor @escaping (Result<Output, Error>) -> Void) {
+    Task {
+      for await result in subscribe() {
+        await onUpdate(result)
+      }
+    }
+  }
+
+  /// Loads the first page of results.
+  /// This method is non-destructive: It will re-fetch the contents of the first page, without modifying any of the other pages, should there be any.
+  public func fetch(cachePolicy: CachePolicy = .returnCacheDataAndFetch) {
+    defer { firstPageWatcher?.refetch(cachePolicy: cachePolicy) }
+    guard firstPageWatcher == nil else { return }
+    self.firstPageWatcher = GraphQLQueryWatcher(
+      client: client,
+      query: initialQuery,
+      resultHandler: { [weak self] result in
+        guard let self else { return }
+        switch result {
+        case .success(let data):
+          let shouldUpdate: Bool
+          if cachePolicy == .returnCacheDataAndFetch && data.source == .cache {
+            shouldUpdate = false
+          } else {
+            shouldUpdate = true
+          }
+
+          self.initialPageResult = data.data
+          guard let firstPageData = data.data else { return }
+          let variables = initialQuery.__variables?.values.compactMap { $0._jsonEncodableValue?._jsonValue } ?? []
+          if shouldUpdate {
+            self.pageOrder.append(variables)
+          }
+          if let latest = self.latest {
+            let (_, nextPage) = latest
+            self.onUpdate?((firstPageData, nextPage, data.source == .cache ? .cache : .fetch))
+          }
+        case .failure(let error):
+          self.onError?(error)
+        }
+      }
+    )
+  }
+
+  /// Loads the next page, based on the latest page info.
+  public func loadMore(
+    cachePolicy: CachePolicy = .fetchIgnoringCacheData
+  ) async throws {
```

Comment:
We could make this return the newly loaded page's data and annotate the function as `@discardableResult` to make the API a little better here.

Reply:
I initially tried that -- but it gave me a lot of pause since the return type here would technically be an `AsyncStream<Output>` given that the watcher can trigger its callback many times (once for a fresh fetch, once for a cached result if it's set to return cached data and fetch, and then again if there's a cache update).